### PR TITLE
Chapter 8: Basic login

### PR DESCRIPTION
**Pull request by phucynwa**

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user && user.authenticate(params[:session][:password])
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: Have you tried removing optional bracket () and see the result ?

File: app/models/user.rb [hunk 2]
```diff
@@ -7,4 +7,9 @@ class User < ApplicationRecord
   validates :email, presence: true, length: {maximum: Settings.max_email_length},
     format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
   validates :password, presence: true, length: {minimum: Settings.min_password_length}
+
+  def User.digest string
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: As there are several ways to declare class method in Ruby, you should declare following way according to the style guide
```
class << self 
  def method_name
    # method body
  end 
end 
```

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (phucynwa) [reply app/controllers/sessions_controller.rb]: @mdmahmudurrahman I tried, it was error.

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<% provide :title, t(".log_in") %>
+<h1><%= t ".log_in" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:session, url: login_path) do |f| %>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: remove ()

File: app/views/sessions/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      <%= f.label :email %>
+      <%= f.email_field :email, class: "form-control" %>
+
+      <%= f.label :password %>
+      <%= f.password_field :password, class: "form-control" %>
+
+      <%= f.submit t(".log_in"), class: "btn btn-primary" %>
+    <% end %>
+
+    <p>New user? <%= link_to t(".sign_up_now"), signup_path %></p>
```
Comment (mdmahmudurrahman) [inline app/views/sessions/new.html.erb]: Use I18n

File: app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+module SessionsHelper
+  def log_in user
+    session[:user_id] = user.id
+  end
+
+  def current_user
+    @current_user ||= User.find_by_id session[:user_id]
```
Comment (mdmahmudurrahman) [inline app/helpers/sessions_helper.rb]: Dynamic is not recommended to use now, you can use find_by and pass the key and value as a parameter

File: app/controllers/sessions_controller.rb (hunk 1 above)
Comment (mdmahmudurrahman) [reply app/controllers/sessions_controller.rb]: You can try with
```
user&.authenticate params[:session][:password]
```

File: app/controllers/sessions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class SessionsController < ApplicationController
+  def new
+  end
+
+  def create
+    user = User.find_by email: params[:session][:email].downcase
+    if user&.authenticate params[:session][:password]
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: Do you know about &. and how it works?

**Pull request merged**